docs/tutorials/gradient-checkpointing.md

```diff
@@ -0,0 +1,488 @@
+
+```{code-cell}
+loss_checkpoint = jax.checkpoint(loss, policy=jax.checkpoint_policies.dots_with_no_batch_dims_saveable)
+print_saved_residuals(loss_checkpoint, params, x, y)
+```
+
+Notice also that by providing a policy, you didn't need to edit the code defining `loss`, `predict`, or `layer`. That is particularly convenient if you want to experiment with policies in calling code (such as a training script) without changing library code (for example, the neural network library).
+
+Some policies can refer to values named with {func}`jax.ad_checkpoint.checkpoint_name`:
+
+```{code-cell}
+from jax.ad_checkpoint import checkpoint_name
+
+def predict(params, x):
+  *Ws, Wlast = params
+  for i, W in enumerate(Ws):
+    x = layer(W, x)
+    x = checkpoint_name(x, name=f'layer{i}_output')
+  x = jnp.dot(Wlast, x)
+  return x
+```
+
+By itself, {func}`jax.ad_checkpoint import.checkpoint_name` is just an identity function. But because some policy functions know to look for them, you can use the names to control whether certain values output by {func}`jax.ad_checkpoint import.checkpoint_name` are considered saveable:
+
+```{code-cell}
+print_saved_residuals(loss, params, x, y)
+```
+
+```{code-cell}
+loss_checkpoint2 = jax.checkpoint(loss, policy=jax.checkpoint_policies.save_any_names_but_these('layer1_output'))
+print_saved_residuals(loss_checkpoint2, params, x, y)
+```
+
+Another policy which refers to names is `jax.checkpoint_policies.save_only_these_names`.
+
+Some of the policies are:
+
+* `everything_saveable`: The default strategy, as if {func}`jax.checkpoint` were not being used at all).
+* `nothing_saveable`: That is, rematerialize everything, as if a custom policy were not being used at all.
+* `dots_saveable`: Or its alias `checkpoint_dots`.
+* `dots_with_no_batch_dims_saveable`: Or its alias `checkpoint_dots_with_no_batch_dims`.
+* `save_anything_but_these_names`: Save any values except for the output of
+  `checkpoint_name` with any of the names given.
+* `save_any_names_but_these`: Save only named values (that is, any outputs of).
+  `checkpoint_name`: Except for those with the names given.
+* `save_only_these_names`: Save only named values, and only among the names
+  given.
+
+Policies only indicate what is saveable; a value is only saved if it's actually needed by the backward pass.
+
+
+#### Advanced: Recursive `jax.checkpoint`
+
+By applying {func}`jax.checkpoint` in the right way, there are many tradeoffs between memory usage and (re)computation that can be expressed. One surprising example is _recursive_ checkpointing, where you apply {func}`jax.checkpoint` to a function which itself calls {func}`jax.checkpoint`-decorated functions in a way so that memory usage from the chain composition of $D$ functions scales like $\mathcal{O}(\log_2 D)$ rather than $\mathcal{O}(D)$.
+
+As a toy example, consider the chain composition of multiple {func}`jnp.sin` functions:
```

Comment:
Change `jnp.sin` to `jax.numpy.sin`. Or alternatively remove `{func}`.

Reply:
SG. Following the `{func}`+`jax.numpy.*` is better.